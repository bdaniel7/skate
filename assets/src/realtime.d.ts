import {
  BlockId,
  DirectionId,
  RouteId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule.d"

import { HeadwaySpacing } from "./models/vehicleStatus"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  scheduledTimepointStatus: VehicleTimepointStatus
}

export type SourceId = string

export type RunId = string

export interface Vehicle {
  id: VehicleId
  label: string
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId
  routeId: RouteId
  tripId: TripId
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string
  operatorName: string
  bearing: number
  blockId: BlockId
  headwaySecs: number | null
  headwaySpacing: HeadwaySpacing | null
  previousVehicleId: string
  scheduleAdherenceSecs: number
  scheduleAdherenceString: string
  scheduledHeadwaySecs: number
  isLayingOver: boolean
  layoverDepartureTime: number | null
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  isOnRoute: boolean
}

export type VehicleOrGhost = Vehicle | Ghost

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId
  stopName: string
}

export interface VehicleScheduledLocation {
  directionId: DirectionId
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}

export interface VehiclesForRoute {
  onRouteVehicles: Vehicle[]
  incomingVehicles: Vehicle[]
  ghosts: Ghost[]
}
