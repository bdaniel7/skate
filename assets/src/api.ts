import { ByRouteId, Route, RouteId, TimepointId } from "./skate.d"

interface RoutesResponse {
  data: Route[]
}

interface TimepointsResponse {
  data: ByRouteId<TimepointId[]>
}

const checkResponseStatus = (response: Response) => {
  if (response.status !== 200) {
    throw new Error(`Response error: ${response.status}`)
  }
  return response
}

const parseJson = (response: Response) => response.json()

export const fetchRoutes = (): Promise<Route[]> =>
  fetch("/api/routes")
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: routes }: RoutesResponse) => routes)
    .catch(error => {
      // tslint:disable-next-line: no-console
      console.error(error)
      throw error
    })

export const fetchTimepointsForRoutes = (
  routeIds: RouteId[]
): Promise<ByRouteId<TimepointId[]>> => {
  const queryParams = routeIds
    .map((routeId: RouteId) => "route_id[]=" + routeId)
    .join("&")
  return fetch(`/api/timepoints?${queryParams}`)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(
      ({ data: timepointsByRouteId }: TimepointsResponse) => timepointsByRouteId
    )
    .catch(error => {
      // tslint:disable-next-line: no-console
      console.error(error)
      throw error
    })
}
