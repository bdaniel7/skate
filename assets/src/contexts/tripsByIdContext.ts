import { createContext } from "react"
import { TripsById } from "../schedule"

const TripsByIdContext = createContext({} as TripsById)

export default TripsByIdContext
