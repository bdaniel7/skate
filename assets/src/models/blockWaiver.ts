import { now } from "../util/dateTime"
import { BlockWaiver } from "../realtime"

export enum CurrentFuturePastType {
  Current = 1,
  Future,
  Past,
}

export const currentFuturePastType = ({
  startTime,
  endTime,
}: BlockWaiver): CurrentFuturePastType => {
  const nowDate: Date = now()

  if (startTime > nowDate) {
    return CurrentFuturePastType.Future
  } else if (endTime < nowDate) {
    return CurrentFuturePastType.Past
  } else {
    return CurrentFuturePastType.Current
  }
}
