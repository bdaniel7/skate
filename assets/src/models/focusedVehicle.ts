import { VehicleId, VehicleOrGhost } from "../realtime"

export interface FocusedVehicle {
  id: VehicleId
  type: FocusType
}

export enum FocusType {
  Selected = 1,
  Highlighted,
}

export const isVehicleSelected = (
  vehicle: VehicleOrGhost,
  focusedVehicle: FocusedVehicle | undefined
): boolean =>
  focusedVehicle !== undefined &&
  focusedVehicle.type === FocusType.Selected &&
  vehicle.id === focusedVehicle.id
