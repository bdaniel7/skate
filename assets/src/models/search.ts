import { Dispatch as ReactDispatch } from "react"
import {
  SearchQuery,
  SavedSearchQuery,
  emptySearchQuery,
  isValidSearchQuery,
} from "./searchQuery"

export interface Search {
  query: SearchQuery
  isActive: boolean
  savedSearches: SavedSearchQuery[]
}

export const initialSearch = {
  query: emptySearchQuery,
  isActive: false,
  savedSearches: [],
}

interface SetSearchTextAction {
  type: "SET_SEARCH_TEXT"
  payload: {
    text: string
  }
}

export const setSearchText = (text: string): SetSearchTextAction => ({
  type: "SET_SEARCH_TEXT",
  payload: { text },
})

interface SetSearchPropertyAction {
  type: "SET_SEARCH_PROPERTY"
  payload: {
    property: string
  }
}

export const setSearchProperty = (
  property: string
): SetSearchPropertyAction => ({
  type: "SET_SEARCH_PROPERTY",
  payload: { property },
})

interface SubmitSearchAction {
  type: "SUBMIT_SEARCH"
}

export const submitSearch = (): SubmitSearchAction => ({
  type: "SUBMIT_SEARCH",
})

export type Action =
  | SetSearchTextAction
  | SetSearchPropertyAction
  | SubmitSearchAction

export type Dispatch = ReactDispatch<Action>

export const reducer = (search: Search, action: Action): Search => {
  switch (action.type) {
    case "SET_SEARCH_TEXT":
      return {
        ...search,
        query: { ...search.query, text: action.payload.text },
        isActive: false,
      }
    case "SET_SEARCH_PROPERTY":
      return {
        ...search,
        query: { ...search.query, property: action.payload.property },
        isActive: false,
      }
    case "SUBMIT_SEARCH":
      if (isValidSearchQuery(search.query)) {
        return {
          ...search,
          isActive: true,
          savedSearches: addSavedSearch(search.savedSearches, {
            text: search.query.text,
          }),
        }
      } else {
        return {
          ...search,
          isActive: false,
        }
      }
  }
  return search
}

export const addSavedSearch = (
  previouslySaved: SavedSearchQuery[],
  newSearch: SavedSearchQuery
): SavedSearchQuery[] =>
  [newSearch]
    .concat(
      previouslySaved.filter(savedSearch => savedSearch.text !== newSearch.text)
    )
    .slice(0, 5)
