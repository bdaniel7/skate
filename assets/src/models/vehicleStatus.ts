import { ScheduleAdherenceStatus, Vehicle } from "../realtime.d"

export type VehicleAdherenceStatus = ScheduleAdherenceStatus | "off-course"

export const status = (vehicle: Vehicle): VehicleAdherenceStatus =>
  vehicle.isOffCourse ? "off-course" : vehicle.scheduleAdherenceStatus
