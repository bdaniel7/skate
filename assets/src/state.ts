import { Dispatch as ReactDispatch } from "react"
import { RouteId } from "./skate.d"

export interface State {
  selectedRouteIds: RouteId[]
}

export const initialState: State = {
  selectedRouteIds: [],
}

interface SelectRoutesAction {
  type: "SELECT_ROUTES"
  payload: {
    routeIds: RouteId[]
  }
}

export const selectRoute = (routeId: RouteId): SelectRoutesAction => ({
  type: "SELECT_ROUTES",
  payload: { routeIds: [routeId] },
})

export const selectRoutes = (routeIds: RouteId[]): SelectRoutesAction => ({
  type: "SELECT_ROUTES",
  payload: { routeIds },
})

interface DeselectRoutesAction {
  type: "DESELECT_ROUTES"
  payload: {
    routeIds: RouteId[]
  }
}

export const deselectRoute = (routeId: RouteId): DeselectRoutesAction => ({
  type: "DESELECT_ROUTES",
  payload: { routeIds: [routeId] },
})

export const deselectRoutes = (routeIds: RouteId[]): DeselectRoutesAction => ({
  type: "DESELECT_ROUTES",
  payload: { routeIds },
})

type Action = SelectRoutesAction | DeselectRoutesAction

export type Dispatch = ReactDispatch<Action>

export const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "SELECT_ROUTES":
      return {
        ...state,
        selectedRouteIds: state.selectedRouteIds.concat(
          action.payload.routeIds
        ),
      }
    case "DESELECT_ROUTES":
      return {
        ...state,
        selectedRouteIds: state.selectedRouteIds.filter(
          id => !action.payload.routeIds.includes(id)
        ),
      }
    default:
      return state
  }
}
