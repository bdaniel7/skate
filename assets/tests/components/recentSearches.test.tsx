import { mount } from "enzyme"
import React from "react"
import renderer from "react-test-renderer"
import RecentSearches from "../../src/components/recentSearches"
import { StateDispatchProvider } from "../../src/contexts/stateDispatchContext"
import { setSearchText } from "../../src/models/search"
import { initialState } from "../../src/state"

describe("RecentSearches", () => {
  test("renders empty state", () => {
    const tree = renderer.create(<RecentSearches />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders with data", () => {
    const searchWithData = {
      text: "999-555",
      property: "run",
      isActive: false,
      savedSearches: [
        { text: "poodle" },
        { text: "999-502" },
        { text: "999-501" },
      ],
    }
    const mockDispatch = jest.fn()
    const tree = renderer
      .create(
        <StateDispatchProvider
          state={{ ...initialState, search: searchWithData }}
          dispatch={mockDispatch}
        >
          <RecentSearches />
        </StateDispatchProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("clicking a recent search sets the search text", () => {
    const searchWithData = {
      text: "999-555",
      property: "run",
      isActive: false,
      savedSearches: [{ text: "poodle" }],
    }
    const mockDispatch = jest.fn()
    const wrapper = mount(
      <StateDispatchProvider
        state={{ ...initialState, search: searchWithData }}
        dispatch={mockDispatch}
      >
        <RecentSearches />
      </StateDispatchProvider>
    )

    wrapper.find(".m-recent-searches__button").simulate("click")

    expect(mockDispatch).toHaveBeenCalledWith(setSearchText("poodle"))
  })
})
