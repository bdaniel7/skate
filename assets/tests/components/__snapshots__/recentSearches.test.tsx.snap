// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentSearches renders empty state 1`] = `
<div
  className="m-recent-searches"
>
  <div
    className="m-recent-searches__heading"
  >
    Recent Searches
  </div>
</div>
`;

exports[`RecentSearches renders with data 1`] = `
<div
  className="m-recent-searches"
>
  <div
    className="m-recent-searches__heading"
  >
    Recent Searches
  </div>
  <button
    className="m-recent-searches__button"
    onClick={[Function]}
  >
    poodle
  </button>
  <button
    className="m-recent-searches__button"
    onClick={[Function]}
  >
    999-502
  </button>
  <button
    className="m-recent-searches__button"
    onClick={[Function]}
  >
    999-501
  </button>
</div>
`;
